wled00/colors.cpp

```diff
@@ -39,21 +39,20 @@ uint32_t color_add(uint32_t c1, uint32_t c2, bool fast)
 {
   if (c1 == BLACK) return c2;
   if (c2 == BLACK) return c1;
-  if (fast) {
-    uint8_t r = R(c1);
-    uint8_t g = G(c1);
-    uint8_t b = B(c1);
-    uint8_t w = W(c1);
-    r = qadd8(r, R(c2));
-    g = qadd8(g, G(c2));
-    b = qadd8(b, B(c2));
-    w = qadd8(w, W(c2));
+  uint32_t rb = (c1 & 0x00FF00FF) + (c2 & 0x00FF00FF); 
```

Comment:
Nice!

Reply:
I thought about this long and hard and then came up with the idea to add (or multiply) two 8bit values in one go on the 32bit system. Also masking two values in one go is faster than shifting twice and then casting to `byte` twice. In "raw" instructions, these changes cut the instructions in half, the compiler may optimize though. Code is still smaller (I checked each of the improvements agains original code in size and they all make code smaller, meaning its also less instructions to execute)

Reply:
I also thought about implementing a new struct that is like CRGB from fastlet, but CRGBW32 colors in 32bit, with `union` access as a native 32bit (would enforce compiler to properly align it in memory too) as well as some casting functions but my C++ knowledge is not at that level to make it good. Doing it like that _may_ improve some passing around of colors, especially with palettes being CRGB (or rewrite the whole fastled stuff to be 32bit native and do away with all the conversions for good but that sounds like a lot of work). The compiler does a good job in making it fast though so not much may be gained in speed.